File: api/bot.py
import os
import logging
import json
import asyncio
from telegram import Update
from telegram.ext import Application, CommandHandler, ContextTypes
import pytz
import aiohttp

# Enable detailed logging
logging.basicConfig(format='%(asctime)s - %(name)s - %(levelname)s - %(message)s', level=logging.INFO)
logger = logging.getLogger(__name__)

TOKEN = os.getenv('TOKEN')
KYIV_TZ = pytz.timezone('Europe/Kyiv')
CHAT_ID = '-4561434244'

async def get_btc_price():
    try:
        async with aiohttp.ClientSession() as session:
            async with session.get('https://api.binance.com/api/v3/ticker/price?symbol=BTCUSDT') as response:
                if response.status == 200:
                    data = await response.json()
                    return float(data['price'])
                else:
                    logger.error(f"API request failed with status code: {response.status}")
                    return None
    except Exception as e:
        logger.error(f"Error fetching BTC price: {str(e)}")
        return None

async def price_command(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    price = await get_btc_price()
    if price:
        await update.message.reply_text(f"Current BTC price: ${price:,.2f}")
    else:
        await update.message.reply_text("Sorry, I couldn't fetch the BTC price at the moment.")

async def webhook(event, context):
    """
    This function processes incoming webhook events from Telegram.
    """
    application = Application.builder().token(TOKEN).build()
    application.add_handler(CommandHandler("price", price_command))

    try:
        update = Update.de_json(json.loads(event['body']), application.bot)
        await application.process_update(update)
    except Exception as e:
        logger.error(f"Error processing update: {str(e)}")

    return {"statusCode": 200}

def handler(event, context):
    """
    This is the main handler function for AWS Lambda / Netlify Function.
    It wraps the asynchronous webhook function in a synchronous call.
    """
    return asyncio.get_event_loop().run_until_complete(webhook(event, context))

File: api/bot.js
const { Telegraf } = require('telegraf');
const axios = require('axios');

const bot = new Telegraf(process.env.TOKEN);
const CHAT_ID = '-1002289271494';

let lastBtcPrice = null;

function formatPrice(price) {
  return price.toLocaleString('en-US', {
    minimumFractionDigits: 3,
    maximumFractionDigits: 3
  });
}

async function getBtcPrice() {
  try {
    console.log('Attempting to fetch BTC price from CoinDesk API');
    const response = await axios.get('https://api.coindesk.com/v1/bpi/currentprice/BTC.json');
    console.log('CoinDesk API response:', response.data);
    return parseFloat(response.data.bpi.USD.rate.replace(',', ''));
  } catch (error) {
    console.error('Error fetching BTC price:', error.message);
    if (error.response) {
      console.error('Error response:', error.response.data);
      console.error('Error status:', error.response.status);
    }
    return null;
  }
}

async function sendBtcPriceUpdate(ctx, price, chatId = CHAT_ID) {
  let message = `🚨 BTC Price Update 🚨\nCurrent BTC price: ${formatPrice(price)} USD`;

  if (lastBtcPrice !== null) {
    const priceChangePercent = ((price - lastBtcPrice) / lastBtcPrice) * 100;
    let emoji;
    if (price > lastBtcPrice) {
      emoji = "🟩";
    } else if (price < lastBtcPrice) {
      emoji = "🔻";
    } else {
      emoji = "▪️";
    }
    message += `\n${emoji} Change: ${priceChangePercent.toFixed(2)}%`;
  }

  await ctx.telegram.sendMessage(chatId, message);
  lastBtcPrice = price;
}

async function checkBtcPrice(ctx) {
  const price = await getBtcPrice();
  if (price) {
    await sendBtcPriceUpdate(ctx, price);
  } else {
    console.error("Failed to fetch BTC price");
  }
}

bot.command('price', async (ctx) => {
  console.log('Price command received');
  const price = await getBtcPrice();
  if (price) {
    await sendBtcPriceUpdate(ctx, price, ctx.chat.id);
  } else {
    await ctx.reply("Sorry, I couldn't fetch the BTC price at the moment.");
  }
});

exports.handler = async (event) => {
  try {
    console.log('Received webhook event:', event.body);
    
    // Check if this is a GitHub Actions trigger
    if (event.body && JSON.parse(event.body).trigger === 'github_action') {
      console.log('Triggered by GitHub Actions');
      await checkBtcPrice(bot);
      return { statusCode: 200, body: 'Price check triggered by GitHub Actions' };
    }
    
    // Handle regular Telegram updates
    await bot.handleUpdate(JSON.parse(event.body));
    
    return { statusCode: 200, body: 'OK' };
  } catch (e) {
    console.error('Error in handler:', e);
    return { statusCode: 400, body: 'This endpoint is meant for bot and telegram communication' };
  }
};